src/services/suggestionDiagnostics.ts

```diff
@@ -109,7 +113,63 @@ namespace ts {
         }
     }
 
+    function addConvertToAsyncFunctionDiagnostics(node: FunctionLikeDeclaration, checker: TypeChecker, diags: DiagnosticWithLocation[]): void {
+
+        const functionType = checker.getTypeAtLocation(node);

```

Comment:
this is an expensive call, since it resolves the return expressions, so you want to move this after the next check to avoid doing the work if it is not needed.